File: 718. Maximum Length of Repeated Subarray/718. Maximum Length of Repeated Subarray top-down.py
from functools import lru_cache
from typing import List


class Solution:
    def findLength(self, nums1: List[int], nums2: List[int]) -> int:
        m, n = len(nums1), len(nums2)
        maxlen = 0

        @lru_cache(None)
        def dp(i: int, j: int) -> int:
            nonlocal maxlen
            if i == m:
                return 0
            if j == n:
                return 0

            dp(i+1, j)
            dp(i, j+1)
            if nums1[i] == nums2[j]:
                res = 1 + dp(i+1, j+1)
                maxlen = max(maxlen, res)
                return res
            else:
                return 0

        dp(0, 0)
        return maxlen


s = Solution()
# nums1 = [1, 2, 3, 2, 1]
# nums2 = [3, 2, 1, 4, 7]
# nums1 = [0, 1, 1, 1, 1]
# nums2 = [1, 0, 1, 0, 1]
nums1 = [0, 0, 0, 0, 0, 0, 1, 0, 0, 0]
nums2 = [0, 0, 0, 0, 0, 0, 0, 1, 0, 0]
print(s.findLength(nums1, nums2))

File: 718. Maximum Length of Repeated Subarray/718. Maximum Length of Repeated Subarray top-down.java
import java.util.Arrays;

class Solution {
    private int m, n, maxlen;
    private int[][] memo;
    private int[] nums1, nums2;

    private int dp(int i, int j) {
        if (i == m)
            return 0;
        if (j == n)
            return 0;

        if (memo[i][j + 1] == -1) {
            memo[i][j + 1] = dp(i, j + 1);
        }
        if (memo[i + 1][j] == -1) {
            memo[i + 1][j] = dp(i + 1, j);
        }
        if (memo[i][j] == -1) {
            if (nums1[i] == nums2[j]) {
                memo[i][j] = 1 + dp(i + 1, j + 1);
                maxlen = Math.max(maxlen, memo[i][j]);
            } else {
                memo[i][j] = 0;
            }
        }
        return memo[i][j];
    }

    public int findLength(int[] nums1, int[] nums2) {
        this.nums1 = nums1;
        this.nums2 = nums2;
        this.m = nums1.length;
        this.n = nums2.length;
        this.maxlen = 0;
        this.memo = new int[m + 1][n + 1];
        for (int[] i : memo) {
            Arrays.fill(i, -1);
        }
        dp(0, 0);
        return maxlen;
    }
}